File: python/product-except-self.py
from typing import List

def productExceptSelf(nums: List[int]) -> List[int]:
    length = len(nums)
    prefix = [1] * length
    suffix = [1] * length
    
    for i in range(1, length):
        prefix[i] = prefix[i - 1] * nums[i - 1]
    for i in range(length - 2, -1, -1):
        suffix[i] = suffix[i + 1] * nums[i + 1]
    
    result = [prefix[i] * suffix[i] for i in range(length)]
    
    return result

if __name__ == "__main__":
    print(productExceptSelf([1,2,3,4]))

File: go/pkg/product_except_self.go
package pkg

func ProductExceptSelf(nums []int) []int {
	len := len(nums)
	prefix := make([]int, len)
	suffix := make([]int, len)
	for i := range nums {
		prefix[i] = 1
		suffix[i] = 1
	}

	for i := 1; i < len; i++ {
		prefix[i] = prefix[i-1] * nums[i-1]
	}
	for i := len - 2; i >= 0; i-- {
		suffix[i] = suffix[i+1] * nums[i+1]
	}

	result := make([]int, len)
	for i := range result {
		result[i] = prefix[i] * suffix[i]
	}

	return result
}
